context/app/static/js/components/workspaces/hooks.ts

```diff
@@ -146,4 +175,33 @@ export function useCreateAndLaunchWorkspace() {
   return { createAndLaunchWorkspace, isCreatingWorkspace };
 }
 
-export { useWorkspacesList, useHasRunningWorkspace, useRunningWorkspace, useLaunchWorkspace };
+function getWorkspaceTimeLeft(workspace: MergedWorkspace) {
+  return findBestJob(workspace?.jobs ?? [])?.job_details?.current_job_details?.time_left;
+}
+
+function useSessionWarning(workspaces: MergedWorkspace[]) {
+  const matchedWorkspace = workspaces.find((workspace) => getWorkspaceTimeLeft(workspace));
+
+  if (!matchedWorkspace) {
+    return false;
+  }
+
+  const timeLeft = getWorkspaceTimeLeft(matchedWorkspace);
+
+  if (!timeLeft) {
+    return false;
+  }
```

Comment:
I think we may be able to avoid the extra call to `getWorkspaceTimeLeft` after a matched workspace is found; what do you think about this approach?

```suggestion
type FoundPair = [MergedWorkspace, number];

function useSessionWarning(workspaces: MergedWorkspace[]) {
  const [matchedWorkspace, timeLeft] = workspaces.reduce<FoundPair | undefined>((acc, ws) => {
  	if (acc) return acc; // Avoids extra calls to `getWorkspaceTimeLeft` after workspace has been found since we can't break a .reduce() 
    const time = getWorkspaceTimeLeft(ws);
    return time ? [ws, time]: acc; // If a time is found, return the workspace with time, otherwise continue iteration
  }, undefined);
  
  if (!matchedWorkspace || !timeLeft) {
  	return false;
  }
```

Reply:
I think this solution is also running into similar issues.